### ISSUE TITLE
Consider skipping derive_partial_eq_without_eq for private types

### ISSUE THREAD

**Issue by djc** [rust-lang/rust-clippy#8875]
### Summary

The new `derive_partial_eq_without_eq` lint seems nice for public types, but I think it might make sense to skip this for types that are not part of a crate's public API. For those types, clearly the `Eq` impl is unneeded, and it would just add more code.

### Reproducer

I tried this code:

```rust
#[derive(Clone, PartialEq)]
enum AstLevel {
    Top,
    Block,
    Nested,
}
```

I expected to see this happen: nothing

Instead, this happened: got `derive_partial_eq_without_eq` warning

### Version

```text
rustc 1.63.0-nightly (b2eed72a6 2022-05-22)
```

**Comment by jonasbb** [comment]
I have this problem in many of my tests. I only derive `Debug, PartialEq` for many types, since that is the only thing needed to make `assert_eq` work. These types are not needed for anything else.

**Comment by jplatte** [comment]
Same for `#[cfg_attr(test, derive(PartialEq))]` on a public type.
